### DISCUSSION TITLE
How to remove a node and not leave an empty new line

### DISCUSSION THREAD

**Discussion by alcroito** [ast-grep/ast-grep#1003] [Q&A] [answered]
With an input file like
```python
# comment 1
a = 1
# comment 2
```

Running  `ast-grep --pattern 'a=1' --rewrite '' --lang python -U`
will remove the assignment but leave the empty line.

```python
# comment 1

# comment 2
```

Is there a way to remove the empty line as well?

**Comment by HerringtonDarkholme** [comment, marked as answer]
whitespace is not encoded in ast so it is impossible to remove the empty line.

The ideal workflow is to apply a formatter after rewriting. 
Note, ast-grep is a code rewrite tool, not a formatter.  https://docs.astral.sh/ruff/formatter/ is recommended.

**Comment by eugenesvk** [comment]
Can Ast-grep expand into Cst-grep to also handle whitespace? (formatter has a much bigger impact that a single rule)

**Reply by HerringtonDarkholme** [reply]
No for current architecture. 

If someone built whitespace-aware tree-sitter, it may be possible.

**Reply by eugenesvk** [reply]
out of curiosity, since tree-sitter is used in text editors that do have to handle whitespace (especially with whitespace-sensitive languages, though I guess in that case whitespace is part of AST?), how does it work there?

**Reply by HerringtonDarkholme** [reply]
tree-sitter does not report whitespace and most tree-sitter apps do not query whitespace as well. A tree-sitter based formatter cannot see a whitespace as TSNode but it can infer the whitespace line by reading surrounding nodes' position.
